cpp/src/rolling/nth_element.cuh

```diff
@@ -0,0 +1,162 @@
+    auto const window_end = window_start + window_size;
+    return n >= 0 ? index_of_nth_non_null(thrust::make_counting_iterator(window_start), window_size)
+                  : index_of_nth_non_null(
+                      thrust::make_reverse_iterator(thrust::make_counting_iterator(window_end)),
+                      window_size);
+  }
+};
+
+/**
+ * @brief Helper function for NTH_ELEMENT window aggregation
+ *
+ * The `N`th element is deemed null for the following cases:
+ *    1. The window has fewer elements than `min_periods`.
+ *    2. N falls outside the window, i.e. N ∉ [-window_size, window_size).
+ *    3. `null_handling == EXCLUDE`, and the window has fewer than `N` non-null elements.
+ *
+ *  If none of the above holds true, the result is non-null. How the value is determined
+ *  depends on `null_handling`:
+ *    1. `null_handling == INCLUDE`: The required value is the `N`th value from the window's start.
+ *    2. `null_handling == EXCLUDE`: The required value is the `N`th *non-null* value from the
+ *        window's start. If the window has fewer than `N` non-null values, the result is null.
+ *
+ * @tparam null_handling Whether to include/exclude null rows in the window
+ * @tparam PrecedingIter Type of iterator for preceding window
+ * @tparam FollowingIter Type of iterator for following window
+ * @param n The index of the element to be returned
+ * @param input The input column
+ * @param preceding Iterator specifying the preceding window bound
+ * @param following Iterator specifying the following window bound
+ * @param min_periods The minimum number of rows required in the window
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return A column the `n`th element of the specified window for each row
+ */
+template <null_policy null_handling, typename PrecedingIter, typename FollowingIter>
+std::unique_ptr<column> nth_element(size_type n,
+                                    column_view const& input,
+                                    PrecedingIter preceding,
+                                    FollowingIter following,
+                                    size_type min_periods,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto const gather_iter = cudf::detail::make_counting_transform_iterator(
+    0,
+    gather_map_constructor<null_handling, PrecedingIter, FollowingIter>{
+      n, input, preceding, following, min_periods, stream, mr});
+  auto gathered = cudf::detail::gather(table_view{{input}},
+                                       gather_iter,
+                                       gather_iter + input.size(),
+                                       cudf::out_of_bounds_policy::NULLIFY,
+                                       stream,
+                                       mr)
```

Comment:
In addition, even in the cases that `gather_iter` is used just once, it may be dereferenced inside a big kernel. This may lead to reduce occupancy/increasing register usage, which ultimately leads to reduced performance. A similar situation is here: https://github.com/rapidsai/cudf/pull/10511